File: pdu/javascript/hako_msgs/pdu_conv_ExecutionUnitRuntimeContext.js
import * as PduUtils from '../pdu_utils.js';
import { ExecutionUnitRuntimeContext } from './pdu_jstype_ExecutionUnitRuntimeContext.js';


/**
 * Deserializes a binary PDU into a ExecutionUnitRuntimeContext object.
 * @param {ArrayBuffer} binary_data
 * @returns { ExecutionUnitRuntimeContext }
 */
export function pduToJs_ExecutionUnitRuntimeContext(binary_data) {
    const js_obj = new ExecutionUnitRuntimeContext();
    const meta_parser = new PduUtils.PduMetaDataParser();
    const meta = meta_parser.load_pdu_meta(binary_data);
    if (meta === null) {
        throw new Error("Invalid PDU binary data: MetaData not found or corrupted");
    }
    binary_read_recursive_ExecutionUnitRuntimeContext(meta, binary_data, js_obj, meta.base_off);
    return js_obj;
}

export function binary_read_recursive_ExecutionUnitRuntimeContext(meta, binary_data, js_obj, base_off) {
    const view = new DataView(binary_data);
    const littleEndian = true;
    // member: config_hash, type: uint32 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 0, 4);
        js_obj.config_hash = PduUtils.binToValue("uint32", bin);
    }
    
    // member: epoch, type: uint8 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 4, 1);
        js_obj.epoch = PduUtils.binToValue("uint8", bin);
    }
    
    // member: owner_id, type: uint8 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 5, 1);
        js_obj.owner_id = PduUtils.binToValue("uint8", bin);
    }
    
    // member: context, type: uint8 (primitive)

    { // varray
        const array_size = view.getInt32(base_off + 8, littleEndian);
        const offset_from_heap = view.getInt32(base_off + 8 + 4, littleEndian);
        const one_elm_size = 1;
        const array_bin = PduUtils.readBinary(binary_data, meta.heap_off + offset_from_heap, one_elm_size * array_size);
        
        if ("uint8" === 'string') {
            js_obj.context = PduUtils.binToValue("string", array_bin);
        } else {
            js_obj.context = PduUtils.binToArrayValues("uint8", array_bin, array_size);
        }
    }
    
    return js_obj;
}

/**
 * Serializes a ExecutionUnitRuntimeContext object into a binary PDU.
 * @param { ExecutionUnitRuntimeContext } js_obj
 * @returns {ArrayBuffer}
 */
export function jsToPdu_ExecutionUnitRuntimeContext(js_obj) {
    const base_allocator = new PduUtils.DynamicAllocator();
    const bw_container = new PduUtils.BinaryWriterContainer(new PduUtils.PduMetaData());

    binary_write_recursive_ExecutionUnitRuntimeContext(0, bw_container, base_allocator, js_obj);

    const base_data_size = base_allocator.size();
    const heap_data_size = bw_container.heap_allocator.size();
    
    bw_container.meta.heap_off = PduUtils.PDU_META_DATA_SIZE + base_data_size;
    bw_container.meta.total_size = bw_container.meta.heap_off + heap_data_size;

    const final_buffer = new ArrayBuffer(bw_container.meta.total_size);
    const final_view = new Uint8Array(final_buffer);

    PduUtils.writeBinary(final_view, 0, bw_container.meta.to_bytes());
    PduUtils.writeBinary(final_view, bw_container.meta.base_off, base_allocator.toArray());
    PduUtils.writeBinary(final_view, bw_container.meta.heap_off, bw_container.heap_allocator.toArray());

    return final_buffer;
}

export function binary_write_recursive_ExecutionUnitRuntimeContext(parent_off, bw_container, allocator, js_obj) {
    const littleEndian = true;
    // member: config_hash, type: uint32 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint32", js_obj.config_hash, 4);
        allocator.add(bin, parent_off + 0);
    }
    
    // member: epoch, type: uint8 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint8", js_obj.epoch, 1);
        allocator.add(bin, parent_off + 4);
    }
    
    // member: owner_id, type: uint8 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint8", js_obj.owner_id, 1);
        allocator.add(bin, parent_off + 5);
    }
    
    // member: context, type: uint8 (primitive)

    { // varray
        const offset_from_heap = bw_container.heap_allocator.size();
        let data_buffer;
        let array_size;
        if ("uint8" === 'string') {
            data_buffer = new TextEncoder().encode(js_obj.context);
            array_size = data_buffer.byteLength;
        } else {
            data_buffer = PduUtils.typesToBin("uint8", js_obj.context);
            array_size = js_obj.context.length;
        }
        bw_container.heap_allocator.add(data_buffer);

        const ref_buffer = new ArrayBuffer(8);
        const ref_view = new DataView(ref_buffer);
        ref_view.setInt32(0, array_size, littleEndian);
        ref_view.setInt32(4, offset_from_heap, littleEndian);
        allocator.add(ref_buffer, parent_off + 8);
    }
    
}

File: pdu/python/hako_msgs/pdu_conv_ExecutionUnitRuntimeContext.py

import struct
from .pdu_pytype_ExecutionUnitRuntimeContext import ExecutionUnitRuntimeContext
from ..pdu_utils import *
from .. import binary_io

# dependencies for the generated Python class



def pdu_to_py_ExecutionUnitRuntimeContext(binary_data: bytearray) -> ExecutionUnitRuntimeContext:
    py_obj = ExecutionUnitRuntimeContext()
    meta_parser = binary_io.PduMetaDataParser()
    meta = meta_parser.load_pdu_meta(binary_data)
    if meta is None:
        raise ValueError("Invalid PDU binary data: MetaData not found or corrupted")
    binary_read_recursive_ExecutionUnitRuntimeContext(meta, binary_data, py_obj, binary_io.PduMetaData.PDU_META_DATA_SIZE)
    return py_obj


def binary_read_recursive_ExecutionUnitRuntimeContext(meta: binary_io.PduMetaData, binary_data: bytearray, py_obj: ExecutionUnitRuntimeContext, base_off: int):
    # array_type: single 
    # data_type: primitive 
    # member_name: config_hash 
    # type_name: uint32 
    # offset: 0 size: 4 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 0, 4)
    py_obj.config_hash = binary_io.binTovalue("uint32", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: epoch 
    # type_name: uint8 
    # offset: 4 size: 1 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 4, 1)
    py_obj.epoch = binary_io.binTovalue("uint8", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: owner_id 
    # type_name: uint8 
    # offset: 5 size: 1 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 5, 1)
    py_obj.owner_id = binary_io.binTovalue("uint8", bin)
    
    # array_type: varray 
    # data_type: primitive 
    # member_name: context 
    # type_name: uint8 
    # offset: 8 size: 1 
    # array_len: 8

    array_size = binary_io.binTovalue("int32", binary_io.readBinary(binary_data, base_off + 8, 4))
    offset_from_heap = binary_io.binTovalue("int32", binary_io.readBinary(binary_data, base_off + 8 + 4, 4))
    one_elm_size = 1 
    array_value = binary_io.readBinary(binary_data, meta.heap_off + offset_from_heap, one_elm_size * array_size)
    py_obj.context = array_value
    
    return py_obj


def py_to_pdu_ExecutionUnitRuntimeContext(py_obj: ExecutionUnitRuntimeContext) -> bytearray:
    binary_data = bytearray()
    base_allocator = DynamicAllocator(False)
    bw_container = BinaryWriterContainer(binary_io.PduMetaData())
    binary_write_recursive_ExecutionUnitRuntimeContext(0, bw_container, base_allocator, py_obj)

    # メタデータの設定
    total_size = base_allocator.size() + bw_container.heap_allocator.size() + binary_io.PduMetaData.PDU_META_DATA_SIZE
    bw_container.meta.total_size = total_size
    bw_container.meta.heap_off = binary_io.PduMetaData.PDU_META_DATA_SIZE + base_allocator.size()

    # binary_data のサイズを total_size に調整
    if len(binary_data) < total_size:
        binary_data.extend(bytearray(total_size - len(binary_data)))
    elif len(binary_data) > total_size:
        del binary_data[total_size:]

    # メタデータをバッファにコピー
    binary_io.writeBinary(binary_data, 0, bw_container.meta.to_bytes())

    # 基本データをバッファにコピー
    binary_io.writeBinary(binary_data, bw_container.meta.base_off, base_allocator.to_array())

    # ヒープデータをバッファにコピー
    binary_io.writeBinary(binary_data, bw_container.meta.heap_off, bw_container.heap_allocator.to_array())

    return binary_data

def binary_write_recursive_ExecutionUnitRuntimeContext(parent_off: int, bw_container: BinaryWriterContainer, allocator, py_obj: ExecutionUnitRuntimeContext):
    # array_type: single 
    # data_type: primitive 
    # member_name: config_hash 
    # type_name: uint32 
    # offset: 0 size: 4 
    # array_len: 1
    type = "uint32"
    off = 0

    
    bin = binary_io.typeTobin(type, py_obj.config_hash)
    bin = get_binary(type, bin, 4)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: epoch 
    # type_name: uint8 
    # offset: 4 size: 1 
    # array_len: 1
    type = "uint8"
    off = 4

    
    bin = binary_io.typeTobin(type, py_obj.epoch)
    bin = get_binary(type, bin, 1)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: owner_id 
    # type_name: uint8 
    # offset: 5 size: 1 
    # array_len: 1
    type = "uint8"
    off = 5

    
    bin = binary_io.typeTobin(type, py_obj.owner_id)
    bin = get_binary(type, bin, 1)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: varray 
    # data_type: primitive 
    # member_name: context 
    # type_name: uint8 
    # offset: 8 size: 1 
    # array_len: 8
    type = "uint8"
    off = 8

    offset_from_heap = bw_container.heap_allocator.size()
    if allocator.is_heap:
        offset_from_heap += 8 # 8 bytes for array_size and offset
    array_size = len(py_obj.context)
    a_b = array_size.to_bytes(4, byteorder='little')
    o_b = offset_from_heap.to_bytes(4, byteorder='little')
    allocator.add(a_b + o_b, expected_offset=parent_off + off)
    binary = binary_io.typeTobin_array(type, py_obj.context, 1)
    bw_container.heap_allocator.add(binary, expected_offset=0)
    

if __name__ == "__main__":
    import sys
    import json

    def print_usage():
        print(f"Usage: python -m pdu.python.pdu_conv_ExecutionUnitRuntimeContext <read|write> [args...]")
        print(f"  read <input_binary_file> <output_json_file>")
        print(f"  write <input_json_file> <output_binary_file>")

    if len(sys.argv) < 2:
        print_usage()
        sys.exit(1)

    command = sys.argv[1]

    if command == "read":
        if len(sys.argv) != 4:
            print_usage()
            sys.exit(1)
        
        binary_filepath = sys.argv[2]
        output_json_filepath = sys.argv[3]

        with open(binary_filepath, "rb") as f:
            binary_data = bytearray(f.read())
        
        py_obj = pdu_to_py_ExecutionUnitRuntimeContext(binary_data)
        
        with open(output_json_filepath, "w") as f:
            f.write(py_obj.to_json())

    elif command == "write":
        if len(sys.argv) != 4:
            print_usage()
            sys.exit(1)

        input_json_filepath = sys.argv[2]
        output_binary_filepath = sys.argv[3]

        with open(input_json_filepath, "r") as f:
            json_str = f.read()
        
        py_obj = ExecutionUnitRuntimeContext.from_json(json_str)
        
        binary_data = py_to_pdu_ExecutionUnitRuntimeContext(py_obj)

        with open(output_binary_filepath, "wb") as f:
            f.write(binary_data)

    else:
        print(f"Unknown command: {command}")
        print_usage()
        sys.exit(1)
